src/middlewared/middlewared/plugins/alert.py

```diff
@@ -0,0 +1,493 @@
+
+POLICIES = ["IMMEDIATELY", "HOURLY", "DAILY", "NEVER"]
+DEFAULT_POLICY = "IMMEDIATELY"
+
+ALERT_SOURCES = {}
+ALERT_SERVICES_FACTORIES = {}
+
+
+class AlertPolicy:
+    def __init__(self, key=lambda now: now):
+        self.key = key
+
+        self.last_key_value = None
+        self.last_key_value_alerts = defaultdict(lambda: defaultdict(dict))
+
+    def receive_alerts(self, now, alerts):
+        gone_alerts = []
+        new_alerts = []
+        key = self.key(now)
+        if key != self.last_key_value:
+            for node in set(alerts.keys()) | set(self.last_key_value_alerts.keys()):
+                for alert_source_name in set(alerts[node].keys()) | set(self.last_key_value_alerts[node].keys()):
+                    for gone_alert in (set(self.last_key_value_alerts[node][alert_source_name].keys()) -
+                                       set(alerts[node][alert_source_name].keys())):
+                        gone_alerts.append(self.last_key_value_alerts[node][alert_source_name][gone_alert])
+                    for new_alert in (set(alerts[node][alert_source_name].keys()) -
+                                      set(self.last_key_value_alerts[node][alert_source_name].keys())):
+                        new_alerts.append(alerts[node][alert_source_name][new_alert])
+
+            self.last_key_value = key
+            self.last_key_value_alerts = copy.deepcopy(alerts)
+
+        return gone_alerts, new_alerts
+
+
+class AlertService(Service):
+    def __init__(self, middleware):
+        super().__init__(middleware)
+
+        self.node = "A"
+
+        self.alerts = defaultdict(lambda: defaultdict(dict))
+
+        self.alert_source_last_run = defaultdict(lambda: datetime.min)
+
+        self.policies = {
+            "IMMEDIATELY": AlertPolicy(),
+            "HOURLY": AlertPolicy(lambda d: (d.date(), d.hour)),
+            "DAILY": AlertPolicy(lambda d: (d.date())),
+            "NEVER": AlertPolicy(lambda d: None),
+        }
+
+    @private
+    async def initialize(self):
+        if not await self.middleware.call("system.is_freenas"):
+            if await self.middleware.call("notifier.failover_node") == "B":
+                self.node = "B"
+
+        for alert in await self.middleware.call("datastore.query", "system.alert"):
+            del alert["id"]
+            alert["level"] = AlertLevel(alert["level"])
+
+            alert = Alert(**alert)
+
+            self.alerts[alert.node][alert.source][alert.key] = alert
+
+        for policy in self.policies.values():
+            policy.receive_alerts(datetime.utcnow(), self.alerts)
+
+        for module in load_modules(os.path.join(os.path.dirname(os.path.realpath(__file__)), os.path.pardir,
+                                                "alert", "source")):
+            for cls in load_classes(module, AlertSource, (FilePresenceAlertSource, ThreadedAlertSource)):
+                source = cls(self.middleware)
+                ALERT_SOURCES[source.name] = source
```

Comment:
There are a few alert modules of legacy system in truenas repo. Perhaps it should allow a custom path to point to truenas alert sources? Or is that planned for later?